### PR TITLE
doc: add refactor-transaction-module design doc

### LINKED ISSUE #4646: Breakdown the "dbms/src/Storages/Transaction" module
## Enhancement
The codes under "dbms/src/Storages/Transaction" are some codes we adapt TiFlash with TiDB. After years of development, it become kind of crowded and confusing.

Suggest breakdown "dbms/src/Storages/Transaction” into these modules:
* Some global context for TiFlash management
* Basic structure from TiDB (TableInfo/DatabaseInfo, etc)
* Syncing schema with TiDB
* Decoding data from row to column (Block)
* Raft layer
  * Region info
  * Handling admin command
  * Handling write command
  * Handling apply snapshot/ingest sst
  * Handling learner read

<hr/>

1. Suggest creating a `dbms/src/TiDB` as the module for basic components from TiDB. And
    * [ ] Rename TMTContext to TiFlashContext
    * [ ] Move TableInfo/DBInfo, etc from TiDB.h to `dbms/src/TiDB`
    * [ ] Move TiFlashContext/ManagedStorages to the `dbms/src/TiDB` module
    * [ ] Add a `dbms/src/TiDB/Codec` and move the logic of decoding data from row to column there
    * [x] Add a `dbms/src/TiDB/Schema` and move schema syncing logic there
2. Rename `dbms/src/Transaction` to `dbms/src/KVStore`. And maybe we should further break down the codes for different Raft processes under this module.
    * [x] Rename `dbms/src/Transaction` to `dbms/src/KVStore` https://github.com/pingcap/tiflash/pull/8073
    * [x] Break down `KVStore`. https://github.com/pingcap/tiflash/pull/8073

### PR DESCRIPTION
**Pull request by hzh0425**
### What problem does this PR solve?

Issue Number: ref #4646 

Problem Summary: refactor the module -- dbms/src/storage/transaction

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: docs/design/2022-05-13-refactor-transaction-module.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
+# Code structure of TiDB and Raft Module
```
Comment (JaySon-Huang) [inline docs/design/2022-05-13-refactor-transaction-module.md:1]: Rename the filename to be "2022-05-13-code-structure-of-tidb-and-raft-module.md"

File: docs/design/2022-05-13-refactor-transaction-module.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+- Author(s): hzh0425
+- Last updated: May. 13, 2022
+- Discussion at: [Breakdown the "dbms/src/Storages/Transaction" module · Issue #4646 · pingcap/tiflash (github.com)](https://github.com/pingcap/tiflash/issues/4646)
+
+# 一 Motivation
```
Comment (JaySon-Huang) [inline docs/design/2022-05-13-refactor-transaction-module.md:7]:
```suggestion
## Background
```

File: docs/design/2022-05-13-refactor-transaction-module.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+- Syncing schema with TiDB
+- Decoding data from row to column (Block)
+- Raft layer
+  - Region info
+  - Handling admin command
+  - Handling write command
+  - Handling apply snapshot/ingest sst
+  - Handling learner read
+
+# 二 Detailed Design
```
Comment (JaySon-Huang) [inline docs/design/2022-05-13-refactor-transaction-module.md:24]:
```suggestion
## Proposal
```

File: docs/design/2022-05-13-refactor-transaction-module.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+## 2.0 Overview
```
Comment (JaySon-Huang) [inline docs/design/2022-05-13-refactor-transaction-module.md:26]:
```suggestion
### Code structure overview
```

File: docs/design/2022-05-13-refactor-transaction-module.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+### Raft Module
+
+When the content in 2.1 is completed, the rest of the components in Transaction are basically related to Raft, so we can rename Transaction to Raft, and add the following folders:
+
+- Raft / Region: Used to place 'Region-related components', such as RegionMeta, RegionData, RegionTable, etc.
+
+- **Do we need to add more folder?**
+
+### Region Module
```
Comment (JaySon-Huang) [inline docs/design/2022-05-13-refactor-transaction-module.md:82]:
```suggestion
#### TiDB Module

First, we need to create a new module -- dbms/src/TIDB to store:

- Some data structures present in transaction/tidb.h: TableInfo / DBInfo etc..
- TiFlashContext and ManagedStorages.
- Module Codec: Decoding data from row to column (Block).
- Module Schema: Syncing schema with TiDB.
- Other components.

#### Codec Module

Codec Module is used to place 'Decoding data from row to column and some serialization/encoding related components', including:

- RowCodec
- JsonCodec
- Datum
  - Datum
  - DatumCodec
- SerializationHelper

#### Schema Module

Schema module is used to place 'Syncing schema with TiDB', including:

- SchemaSyncer & TiDBSchemaSyncer
- SchemaSyncSerivce
- SchemaBuilder
- SchemaGetter
- SchemaNameMapper

#### Raft Module

When the content in 2.1 is completed, the rest of the components in Transaction are basically related to Raft, so we can rename Transaction to Raft, and add the following folders:

- Raft / Region: Used to place 'Region-related components', such as RegionMeta, RegionData, RegionTable, etc.

- **Do we need to add more folder?**

#### Region Module
```

Comment (hzh0425) [reply docs/design/2022-05-13-refactor-transaction-module.md]: thanks a lot

File: docs/design/2022-05-13-code-structure-of-tidb-and-raft-module.md.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  - Handling apply snapshot/ingest sst
+  - Handling learner read
+
+## Proposal
+
+### Code structure overview
+
+```
+dbms / src
+------ TIDB
```
Comment (JaySon-Huang) [inline docs/design/2022-05-13-code-structure-of-tidb-and-raft-module.md.md:30]:
```suggestion
------ TiDB
```

File: docs/design/2022-05-13-code-structure-of-tidb-and-raft-module.md.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+- SchemaSyncSerivce
+- SchemaBuilder
+- SchemaGetter
+- SchemaNameMapper
+
+#### Raft Module
+
+When the previous module is completed, the rest of the components in Transaction are basically related to Raft, so we can rename Transaction to Raft, and add the following folders:
+- Raft / Region: Used to place 'Region-related components', such as RegionMeta, RegionData, RegionTable, etc.
+- **Do we need to add more folder?**
```
Comment (JaySon-Huang) [inline docs/design/2022-05-13-code-structure-of-tidb-and-raft-module.md.md:71]:
```suggestion
```

**Pull request closed without merging**